### PR TITLE
Add documentation for qute support

### LINKED ISSUE #427: Add documentation for qute support
Like we did for vscode microprofile and xml we need to have documentation for showing in action our qute support.

We should have 2 entries

 * qute support indsde java file

We must explain that checkedtemplate and template field are supported to manage link with qute template by providing codelens diagnostics and document link.

 * qute support inside template file

We have completion for parameter declaration sections and java data model.

We have codelens which shows data model coing from checked template and template field.

We have validation to validate qute template syntax and java data model

We have definition to go to the declared variable to template or to java

We have hover to show java type and doc for redolver

We have linked editing range to update variable

We have higlight to higligt variable and referenced variable . To highlight start ebd sectio

### PR DESCRIPTION
**Pull request by angelozerr**
Add documentation for qute support

Fixes #427

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,11 +5,17 @@
...
 
-This Visual Studio Code extension provides support for [Quarkus](https://quarkus.io/) and [MicroProfile](https://github.com/eclipse/microprofile) development
+This Visual Studio Code extension provides support for :
+
+ * [Quarkus](https://quarkus.io/) and [MicroProfile](https://github.com/eclipse/microprofile) development
 by extending [Visual Studio Code extension for MicroProfile](https://github.com/redhat-developer/vscode-microprofile) with Quarkus features.
 
 ![](images/propertiesSupport.png)
 
+ * [Qute support](docs/qute/README.md) to provide completion, validation, etc for the [Qute template engine](https://quarkus.io/guides/qute-reference) inside Java and template.
```
Comment (fbricon) [inline README.md]: etc.

Comment (fbricon) [reply README.md]: inside Java and template.=> in Java code and Qute templates

File: docs/qute/JavaSupport.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Qute Java support
+
+Qute support provides `validation`, `documentation link`, `codelens` inside Java file to manage binding with a Qute template:
```
Comment (fbricon) [inline docs/qute/JavaSupport.md]: Java files,  to manage binding with Qute templates:

File: docs/qute/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Qute support
+
+[Qute](https://quarkus.io/guides/qute-reference) is the template engine used in Quarkus application.
```
Comment (fbricon) [inline docs/qute/README.md]: is a template engine used in Quarkus applications.

File: docs/qute/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Qute support
+
+[Qute](https://quarkus.io/guides/qute-reference) is the template engine used in Quarkus application.
+
+[vscode-quarkus](https://github.com/redhat-developer/vscode-quarkus) provides a `Qute support` inside:
```
Comment (fbricon) [inline docs/qute/README.md]: provides `Qute` support in:

File: docs/qute/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Qute support
+
+[Qute](https://quarkus.io/guides/qute-reference) is the template engine used in Quarkus application.
+
+[vscode-quarkus](https://github.com/redhat-developer/vscode-quarkus) provides a `Qute support` inside:
+
+ * [template files](TemplateSupport.md#template) : completion, validation etc for Qute template.
```
Comment (fbricon) [inline docs/qute/README.md]: etc. for Qute templates

File: docs/qute/README.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# Qute support
+
+[Qute](https://quarkus.io/guides/qute-reference) is the template engine used in Quarkus application.
+
+[vscode-quarkus](https://github.com/redhat-developer/vscode-quarkus) provides a `Qute support` inside:
+
+ * [template files](TemplateSupport.md#template) : completion, validation etc for Qute template.
+ * [Java files](JavaSupport.md#java-file) : completion, validation etc for the Qute Quarkus integration.
```
Comment (fbricon) [inline docs/qute/README.md]: etc. for Qute integration within Quarkus applications

File: docs/qute/TemplateSupport.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
+# Qute Template support
+
+Qute support provides `completions`, `validation`, `hover`, `definition`, `highlight`, `linked editing range` inside Qute template.
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: `linked editing` in Qute templates

File: docs/qute/TemplateSupport.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
+# Qute Template support
+
+Qute support provides `completions`, `validation`, `hover`, `definition`, `highlight`, `linked editing range` inside Qute template.
+
+To benefit with this support, we recommand to name your template file with the `*.qute.*` extension (ex: `items.qute.html`, `items.qute.txt`, `items.qute.yaml`, etc)
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: we recommend your template files to follow the `*.qute.*` naming pattern (ex: `items.qute.html`, `items.qute.txt`, `items.qute.yaml`, etc.).

File: docs/qute/TemplateSupport.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
+# Qute Template support
+
+Qute support provides `completions`, `validation`, `hover`, `definition`, `highlight`, `linked editing range` inside Qute template.
+
+To benefit with this support, we recommand to name your template file with the `*.qute.*` extension (ex: `items.qute.html`, `items.qute.txt`, `items.qute.yaml`, etc)
+
+## Parameter declaration support
+
+A `parameter declaration` is used to declare the expected model in the Qute template:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: in a Qute template

File: docs/qute/TemplateSupport.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+## Parameter declaration support
+
+A `parameter declaration` is used to declare the expected model in the Qute template:
+
+```
+{@java.util.List<org.acme.qute.Item> items}
+```
+
+The Qute support helps to write easily this declaration with the completion, hover and definition support.
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: to easily write

File: docs/qute/TemplateSupport.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+### Completion for Java types
+
+The Qute support provides completion to write easily parameter declaration:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: to easily write

File: docs/qute/TemplateSupport.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+![Completion for Java types](images/QuteTemplateParameterDeclarationCompletion.gif)
+
+### Validate Java type
+
+The Qute support validate Java types:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: validates

File: docs/qute/TemplateSupport.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+You can jump to the proper Java type:
+
+![Jump to Java type definition](images/QuteTemplateParameterDeclarationDefinition.gif)
+
+## Section support
+
+### Completion for Qute section
+
+Qute support provides snippet to generate quickly `Qute section`. Here a sample to generate quickly the `#for` section:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: provides snippets to quickly generate `Qute` sections. Here is a sample for the `#for` section:

File: docs/qute/TemplateSupport.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+![Completion for Qute section](images/QuteTemplateSectionSnippetCompletion.gif)
+
+### Linked editing range
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: Linked editing

File: docs/qute/TemplateSupport.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+Linked editing is supported, allowing for simultaneous changes an object part expression.
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: allowing for simultaneous renaming of a parameter expression.

File: docs/qute/TemplateSupport.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+To enable this feature, the setting `editor.linkedEditing` must be set to `true` in your `settings.json`.
+
+![Linked EditingRange](images/QuteTemplateLinkedEditingRange.gif)
+
+## Tag Name Highlighting
+
+When placing the cursor on an Qute section tag, the corresponding opening/closing tag will be highlighted with their nested block:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: on a Qute section tag, the corresponding opening/closing tag will be highlighted with their nested block:

File: README.md (hunk 1 above)
Comment (angelozerr) [reply README.md]: Sorry I don't understand, you want I add a `.` after etc?

File: docs/qute/README.md [hunk 17]
```diff
@@ -0,0 +1,12 @@
+# Qute support
+
+[Qute](https://quarkus.io/guides/qute-reference) is the template engine used in Quarkus application.
+
+[vscode-quarkus](https://github.com/redhat-developer/vscode-quarkus) provides a `Qute support` inside:
+
+ * [template files](TemplateSupport.md#template) : completion, validation etc. for Qute template.
```
Comment (fbricon) [inline docs/qute/README.md]: templates

File: docs/qute/TemplateSupport.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
+# Qute Template support
+
+Qute support provides `completions`, `validation`, `hover`, `definition`, `highlight`, `linked editing range` inside Qute template.
+
+To benefit with this support, we recommand to name your template file with the `*.qute.*` extension (ex: `items.qute.html`, `items.qute.txt`, `items.qute.yaml`, etc.)
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: we recommend your template files to follow the `*.qute.*` naming pattern (ex: `items.qute.html`, `items.qute.txt`, `items.qute.yaml`, etc.).

File: docs/qute/TemplateSupport.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+### Completion for Java data model
+
+Completion is available for allowed Java `fields`, `methods` and `virtual methods` (coming from the Qute core like `orEmpty` or coming from @TemplateExtension):
+
+![Completion for Java data model](images/QuteTemplateCompletionForJavaDataModel.gif)
+
+### Validation for Java data model
+
+Validation highlight as an error the unkwown Java field, method:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: highlights unknown Java fields or methods as an error:

File: docs/qute/TemplateSupport.md [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+
+![Validation for Java data model](images/QuteTemplateValidationForJavaDataModel.png)
+
+### Hover for Java data model
+
+When you hover object, property, method part, you can see the Java type:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: hover over an object, property or method,

File: README.md (hunk 1 above)
Comment (angelozerr) [reply README.md]: fixed

File: docs/qute/TemplateSupport.md [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+![Hover for Java data model](images/QuteTemplateCompletionForJavaDataModel.gif)
+
+### Jump to Java field, method definition
+
+`Ctrl+Click` on property, method part jump to the definition of the Java field, method or virtual method (for @TemplateExtension):
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: on a property or method to jump to the definition of the matching Java field, method or virtual method

File: docs/qute/TemplateSupport.md [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+![Definition for Java data model](images/QuteTemplateDefinitionForJavaDataModel.gif)
+
+## Quarkus integration for Qute
+
+When the template is used in a Quarkus application and it is bound with a Java Template field or a method with @CheckedTemplate, you can see the expected model as CodeLens:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: method annotated with `@CheckedTemplate`

File: docs/qute/TemplateSupport.md [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+![CodeLens for Java data model](images/QuteTemplateCodeLensJavaDataModel.png)
+
+In this case, you don't need to declare parameter declaration to benefit with Qute completion, validation, etc. Java data model.
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: of the Java data model.

File: docs/qute/TemplateSupport.md [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+
+CodeLens are clickable to jump to the proper Java type, method field.
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: method or field.

**Review by fbricon — CHANGES_REQUESTED** [review]

File: README.md [hunk 25]
```diff
@@ -5,11 +5,17 @@
...
 
-This Visual Studio Code extension provides support for [Quarkus](https://quarkus.io/) and [MicroProfile](https://github.com/eclipse/microprofile) development
+This Visual Studio Code extension provides support for :
+
+ * [Quarkus](https://quarkus.io/) and [MicroProfile](https://github.com/eclipse/microprofile) development
 by extending [Visual Studio Code extension for MicroProfile](https://github.com/redhat-developer/vscode-microprofile) with Quarkus features.
 
 ![](images/propertiesSupport.png)
 
+ * [Qute support](docs/qute/README.md) to provide completion, validation, etc. for the [Qute template engine](https://quarkus.io/guides/qute-reference) inside Java and template.
```
Comment (AlexXuChen) [inline README.md]: "...Java and template files."

File: docs/qute/TemplateSupport.md (hunk 24 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 23 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 22 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 21 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 20 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 19 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 18 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/README.md (hunk 17 above)
Comment (angelozerr) [reply docs/qute/README.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 16 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 15 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 14 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 13 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 12 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 11 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 10 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 9 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 8 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/README.md (hunk 6 above)
Comment (angelozerr) [reply docs/qute/README.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 7 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/README.md (hunk 4 above)
Comment (angelozerr) [reply docs/qute/README.md]: fixed

File: docs/qute/README.md (hunk 5 above)
Comment (angelozerr) [reply docs/qute/README.md]: fixed

File: docs/qute/JavaSupport.md (hunk 2 above)
Comment (angelozerr) [reply docs/qute/JavaSupport.md]: fixed

File: docs/qute/README.md (hunk 3 above)
Comment (angelozerr) [reply docs/qute/README.md]: fixed

File: docs/qute/TemplateSupport.md [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+
+Qute support provides `completions`, `validation`, `hover`, `definition`, `highlight`, `linked editing` in Qute template.
+
+To benefit with this support, we recommend your template files to follow the `*.qute.*` naming pattern (ex: `items.qute.html`, `items.qute.txt`, `items.qute.yaml`, etc.).
+
+## Parameter declaration support
+
+A `parameter declaration` is used to declare the expected model in a Qute template:
+
+```
```
Comment (AlexXuChen) [inline docs/qute/TemplateSupport.md:11]: maybe add 'html' here, and other qute snippets to show same syntax highlighting

File: docs/qute/TemplateSupport.md [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+
+`Ctrl+Click` on a property or method to jump to the definition of the matching Java field, method or virtual method (for `@TemplateExtension`):
+
+![Definition for Java data model](images/QuteTemplateDefinitionForJavaDataModel.gif)
+
+## Quarkus integration for Qute
+
+When the template is used in a Quarkus application and it is bound with a Java Template field or a method annotated with `@CheckedTemplate`, you can see the expected model as CodeLens:
+
+![CodeLens for Java data model](images/QuteTemplateCodeLensJavaDataModel.png)
```
Comment (AlexXuChen) [inline docs/qute/TemplateSupport.md:101]: Maybe clean this image up a bit and use straight lines/arrows rather then freehand

File: docs/qute/TemplateSupport.md [hunk 28]
```diff
@@ -0,0 +1,93 @@
...
+
+In this case, you don't need to declare parameter declaration to benefit with Qute completion, validation, etc. of the Java data model.
```
Comment (AlexXuChen) [inline docs/qute/TemplateSupport.md]: "benefit from..."

**Review by AlexXuChen — CHANGES_REQUESTED** [review]

File: docs/qute/TemplateSupport.md (hunk 26 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md:11]: The parameter declaration is a Qute syntax not an html syntax, so we cannot higlight Qute syntax, but perhaps it should be really interesting to reference our Qute syntax inside github. It seems that we should create a PR at https://github.com/github/linguist @fbricon what do you think about this idea?

File: README.md (hunk 25 above)
Comment (angelozerr) [reply README.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 27 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md:101]: fixed

File: docs/qute/TemplateSupport.md (hunk 28 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 26 above)
Comment (fbricon) [reply docs/qute/TemplateSupport.md:11]: +1

File: docs/qute/TemplateSupport.md [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+
+![Validation for Java data model](images/QuteTemplateValidationForJavaDataModel.png)
+
+### Hover for Java data model
+
+When you hover over an object, property or method or virtual method (for `@TemplateExtension`), you can see the Java type:
+
+![Hover for Java data model](images/QuteTemplateHoverForJavaDataModel.gif)
+
+When you hover over a [Built-in Resolvers](https://quarkus.io/guides/qute-reference#built-in-resolvers) you can see relevant documentation with a link to the documentation:
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: ... you can see the relevant documentation, with a link to the Qute reference guide.

File: docs/qute/TemplateSupport.md [hunk 30]
```diff
@@ -0,0 +1,105 @@
...
+
+### Jump to Java field, method definition
+
+`Ctrl+Click` on a property or method to jump to the definition of the matching Java field, method or virtual method (for `@TemplateExtension`):
+
+![Definition for Java data model](images/QuteTemplateDefinitionForJavaDataModel.gif)
+
+### Jump to the declared object
+
+A variable can be declared inside a Qute template (with parameter declaration, or #let section) or declared from Java side.
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: from the Java side.

Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md (hunk 29 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

File: docs/qute/TemplateSupport.md [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+
+`Ctrl+Click` on a property or method to jump to the definition of the matching Java field, method or virtual method (for `@TemplateExtension`):
+
+![Definition for Java data model](images/QuteTemplateDefinitionForJavaDataModel.gif)
+
+### Jump to the declared object
+
+A variable can be declared inside a Qute template (with parameter declaration, or #let section) or declared from the Java side.
+
+`Ctrl+Click` on a object part jump to the declared object in Qute template (with parameter declaration, or #let section) or in Java side (method parameter or data method invocation):
```
Comment (fbricon) [inline docs/qute/TemplateSupport.md]: on an object part jumps to the declared object in the Qute template (...), or to the Java code (...):

**Review by fbricon — CHANGES_REQUESTED** [review]

File: docs/qute/TemplateSupport.md (hunk 31 above)
Comment (angelozerr) [reply docs/qute/TemplateSupport.md]: fixed

**Pull request merged**